### PR TITLE
Version 2.0.0

### LINKED ISSUE #7: Better statistics
# Feature
Improve the commands that display statistics and information about the bot.
Additionally, have a command that displays the statistics of the server that it's executed in.

## Acceptance Criteria
- [x] Sushi Bot's general statistics command will be improved.
- [x] Sushi Bot's information command will be improved.
- [x] Sushi Bot may receive other improvements.
- [x] Sushi Bot shall have a command for displaying the statistics of the server the command is executed in.
- [x] The statistics provided by the command shall include the following:
    - [x] The amount of users in the server,
    - [x] The amount of users in the server who are currently online, and
    - [x] The amount of bots in the server.

## Additional Context
The server stats command will be ported from Sushi Stats and made into slash commands.

### LINKED ISSUE #34: Welcome and Goodbye messages
# Feature Request
A feature were Sushi bot will welcome all new members who join a server and say goodbye to them when they leave.

## Acceptance Criteria
- [x] Sushi bot should have a default message for welcoming new members and for saying goodbye when they leave the server.
  - [x] Users should be able to write their own welcome/goodbye message for Sushi bot to use instead.
  - [x] Users should be able to revert to the default message.
- [x] Users should be able to select which text channel Sushi bot should post the welcome/goodbye messages in.

### LINKED ISSUE #35: Action commands enhancements
# Feature
Various enhancements for the action commands, such as differing responses depending on action target.

## Acceptance Criteria
- [x] The action commands should have a different response when the target is the user themselves.
- [x] The action commands should have a different response when the target is Sushi Bot.
- [x] The responses should fit the action commands.
- [x] The action commands may be cleaned up in various places.

### LINKED ISSUE #36: Documentation for v2.0.0
# Task
Prepare the repository's documentation for the v2.0.0 release. This will include the README and wiki pages.

## Task Definitions
- [x] Finish the parts in `README.md` marked with TODO.
- [x] Add sections in the server configuration wiki page for the welcome- and goodbye features.

## Additional Context
More tasks may be added if the need arises.

### PR DESCRIPTION
**Pull request by MirageAegis**
# What Does This PR Do?
This PR brings automatic welcome- and goodbye messages to Sushi Bot, which will be the final additions for version 2.0.0.
This means that Sushi Bot v2.0.0 will be released once this PR has been merged.

This PR also adds missing documentation to the README and wikis.

## Acceptance Criteria Met
- Sushi Bot's general statistics command has been improved.
- Sushi Bot's information command has been changed to an invite command and been improved.
- Sushi Bot has a command for displaying server statistics of the server the command is executed in, including:
  - The amount of users in the server,
  - The amount of users in the server who are currently online,
  - The amount of users in the server who are currently offline, and
  - The amount of bots in the server.

- Sushi Bot has configurable automatic welcome messages.
- Sushi Bot has configurable automatic goodbye messages.
  - Both of these features have default messages.

- The action commands have different responses depending on who the target is.
- The action commands have been cleaned up a little.

## Task(s) Completed
- Finished the TODOs in `README.md`.
- Added sections for the new features in the server configuration wiki page.

## Related Issues
Closes #7 
Closes #34 
Closes #35 
Closes #36

### CODE REVIEW

File: src/commands/configuration/config-subcommands/welcome.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                value: "Optional parameter. The message to greet with. Omit to use the default message\n" +
+                       "Variables:\n" +
+                       `- \`${NAME}\` - Replaced with the new member's display name\n` +
+                       `- \`${PING}\` - Replaced with a mention to the new member\n` +
+                       `- \`${SERVER}\` - Replaced with the server's name\n` +
+                       `- \`${OWNER}\` - Replaced with the server owner's display name\n` +
+                       `- \`${NEW_LINE}\` - Replaced with a new line`
+            }
+        )
+};
```
Comment (MirageAegis) [inline src/commands/configuration/config-subcommands/welcome.ts:106]: This is the `/config welcome` commands. It and `/config goodbye` are quite similar to `/config golive` where a server administrator provides a channel to use and optionally a template message.

File: src/commands/action/bonk.ts [hunk 2]
```diff
@@ -35,32 +38,64 @@ import { TenorSingleton } from "../../util/tenor-utils.js";
...
+            case bonked.id === ctx.client.user.id:
+                // User tries to bonk Sushi Bot
+                response = "Oh no you don't\n" +
+                           `${ctx.client.user} bonks ${bonker}!`;
+                break;
+            default:
+                // User bonks someone else
+                response = `${bonker} bonks ${bonked}. Ouchies!`;
+                break;
+        }
```
Comment (MirageAegis) [inline src/commands/action/bonk.ts:83]: This section has been added to all action commands, to make Sushi Bot respond differently when the target is the user themselves, or if the target is Sushi Bot.

File: src/events/greet.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    const displayName: string = member.displayName;
+    const server: Guild = member.guild;
+    const owner: string = (await server.fetchOwner()).displayName;
+
+    return template.replace(NAME_RE, displayName)
+        .replace(PING_RE, `${member}`)
+        .replace(SERVER_RE, server.name)
+        .replace(OWNER_RE, owner)
+        .replace(NEW_LINE_RE, "\n");
+};
```
Comment (MirageAegis) [inline src/events/greet.ts:117]: This is the string formatter for welcome- and goodbye messages. It's similar to the one used to auto go-live posts- and shout outs.

File: src/events/events.ts [hunk 4]
```diff
@@ -106,6 +107,56 @@ const logTo = async (
...
+                break;
+            case Action.Goodbye:
+                server.goodbye = null;
+                break;
+        }
+        await server.save();
+    }
+
+    await channel.send(await formatGreeting(member, action, greeting.message));
+};
```
Comment (MirageAegis) [inline src/events/events.ts:158]: This is the helper function that welcomes new members says goodbye to members who leave. It stays silent when there aren't any configurations, or if the member in question is a bot.

File: src/events/events.ts [hunk 5]
```diff
@@ -130,7 +183,17 @@ export const onMemberJoin = async (client: Client, member: GuildMember): Promise
...
         });
+        return;
     }
+
+    // ----- !CHECK BLACKLIST -----
+
+
+    // ----- GREET MEMBERS -----
+
+    await greet(client, member, server, Action.Welcome);
```
Comment (MirageAegis) [inline src/events/events.ts:194]: Sushi Bot only welcomes users who aren't blacklisted.

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "sushi-bot",
-    "version": "2.0.0-alpha",
+    "version": "2.0.0",
```
Comment (MirageAegis) [inline package.json:3]: Finally, the version has been changed from `2.0.0-alpha` to `2.0.0`

**Review by MirageAegis — COMMENTED** [review]
This review will walk through all noteworthy changes in this PR.

**Review by zahatikoff — APPROVED** [review]

**Pull request merged**